:memo: Docs refactor (#570)
sequinstream/sequin: docs/how-to/stream-postgres-to-a-webhook-endpoint.mdx

---
title: "How to stream Postgres to a webhook endpoint"
sidebarTitle: "Stream to webhook endpoint"
description: "Receive Postgres changes as webhooks in real-time"
---

> This guide shows you how to use Sequin to set up real-time streaming of database changes from Postgres to your webhook endpoints.

Reviewer: I wonder if we need to anchor on CDC more to echo the value prop:

```suggestion
This guide shows you how to use Sequin to set up Postgres CDC to stream database changes to your webhook endpoints.
```

With Postgres data streaming to your application, you can trigger workflows, keep services in sync, [build audit logs](/how-to/create-audit-logs), [maintain caches](/how-to/maintain-caches), and more.

By the end of this how-to, you'll have a configured HTTP endpoint receiving database changes.

## Prerequisites

This guide assumes you:

> 1. [Already have Sequin installed](/quickstart/webhooks)

Reviewer 2: is this the right link for installation?

Author: @RTLS I'm thinking for every how-to, the pre-requisite will be a quickstart for the relevant sink

Your webhook endpoint must return a `200 OK` after successfully processing a batch of messages. Sequin will consider any other response code a failure, and will [mark the message for re-delivery](reference/sinks/webhooks#delivery-configuration).

<Tip>
  If you're using Sequin Cloud and developing locally, you can use the Sequin CLI's [`tunnel` command](/reference/cli/overview#tunnel) to connect Sequin's servers to a local HTTP endpoint. So, no need to deploy your application just yet!
</Tip>

<Tip>
  If you're not ready to add an HTTP endpoint to your application just yet, you can use a service like [Webhook.site](https://webhook.site) to receive webhooks.
</Tip>

#### Authentication

Before deploying your HTTP endpoint to production, you should add authentication. This way, your application can verify that requests are coming from Sequin.

A simple strategy is to generate a secure token and require it in the `Authorization` header of inbound requests. You'll supply this token to Sequin by adding it as a header for your HTTP endpoint.

Here's how to generate a secure token with `openssl`:

```bash
openssl rand -base64 32
```

> ### (Optional) Create a Change Capture Pipeline

Reviewer: The big value prop here is retention on changes. So maybe the headline is "(Optional) Enable retention for changes" or something like that. And then the section highlight when you might want to do that and how (using the change capture pipeline)

  </Step>

  <Step title="Choose message type">
    Specify whether you want to receive [changes or rows](/reference/messages) from the table.

    If you're not sure which to choose, select **Changes**. Changes are a better fit for event-driven use cases, like triggering workflows, which is usually what you want to do with webhooks.
  </Step>

  <Step title="Specify filters">
    If you selected changes, in "Records to process", you can indicate whether you want to receive `insert`, `update`, and/or `delete` changes.

    You can also specify [SQL filters](/reference/filters) to narrow down the webhooks you want to receive. For example, if you only want to receive webhooks for `subscriptions` that currently have an `mrr` greater than $100, you can add a filter on `mrr > 100`.
  </Step>

  <Step title="Specify backfill">
    You can optionally indicate if you want your webhook endpoint to receive a [backfill](reference/backfills) of all or a portion of the table's existing data. Backfills are useful if you want your webhook endpoint to process historical data. For example, if you're materializing a cache, you might want to warm it with existing rows.

>     You can backfill at any time. If you don't want to backfill, toggle "Backfill" off.

Reviewer 2: `backfill` terminology seems correct to me

    Conversely, for replication use cases, like batch upserting to a database, you'll want to set this to an optimal value for that batch upsert operation.
  </Step>

  <Step title="Specify advanced delivery configuration">
    See the [webhook sink reference](/reference/sinks/webhooks#delivery-configuration) for more information on advanced delivery configuration options.
  </Step>

  <Step title="Select the HTTP endpoint you created">
    Under "HTTP Endpoint", select the HTTP endpoint you created earlier.
  </Step>

  <Step title="Create the sink">
    Give your sink a name, then click "Create Webhook Sink".
  </Step>
</Steps>

## Verify & debug

To verify that your webhook sink is working:

1. Make some changes in your source table.
2. Verify that the count of messages for your sink increases in the Sequin web console.
3. Verify receipt of messages in your webhook endpoint logs.

If messages don't seem to be flowing:

1. Click the "Messages" tab to view the state of messages for your sink.
2. Click any failed message.

> 3. Check the delivery logs for error details, including the response from your webhook endpoint.

Reviewer 2: yep we need to add responses here!
